R/fread.R

```diff
@@ -142,6 +148,136 @@ fread <- function(input="",file=NULL,text=NULL,cmd=NULL,sep="auto",sep2="auto",d
+    if (is.character(skip))
+      warning("Combining a search string as 'skip' and reading a ",
+              "YAML header may not work as expected -- currently, ",
+              "reading will proceed to search for 'skip' from ",
+              "the beginning of the file, NOT from the end of ",
+              "the metadata; please file an issue on GitHub if ",
+              "you'd like to see more intuitive behavior supported.")
+    # create connection to stream header lines from file:
+    #   https://stackoverflow.com/questions/9871307
+    f = base::file(input, 'r')
+    first_line = readLines(f, n=1L)
+    n_read = 1L
+    yaml_border_re = '^#?---'
+    if (!grepl(yaml_border_re, first_line)) {
+      close(f)
+      stop('Encountered <', substring(first_line, 1L, 50L),
+           if (nchar(first_line) > 50L) '...', '> at the first ',
+           'unskipped line (', 1L+skip, '), which does not ',
+           'constitute the start to a valid YAML header ',
+           '(expecting something matching regex "', yaml_border_re,
+           '"); please check your input and try again.')
+    }
+
+    yaml_comment_re = '^#'
+    yaml_string = character(0L)
+    while (TRUE) {
+      this_line = readLines(f, n=1L)
+      n_read = n_read + 1L
+      if (!length(this_line)){
+        close(f)
+        stop('Reached the end of the file before finding ',
+             'a completion to the YAML header. A valid ',
+             'YAML header is bookended by lines matching ',
+             'the regex "', yaml_border_re, '". Please ',
+             'double check the input file is a valid csvy.')
+      }
+      if (grepl(yaml_border_re, this_line)) break
+      if (grepl(yaml_comment_re, this_line))
+        this_line = sub(yaml_comment_re, '', this_line)
+      yaml_string = paste(yaml_string, this_line, sep='\n')
+    }
+    close(f) # when #561 is implemented, no need to close f.
+
+    yaml_header = yaml::yaml.load(yaml_string)
+    yaml_names = names(yaml_header)
+    if (verbose) cat('Processed', n_read, 'lines of YAML',
+                     'metadata with the following top-level fields:',
+                     brackify(yaml_names), '\n')
+    # process header first since it impacts how to handle colClasses
+    if ('header' %chin% yaml_names) {
+      if ('header' %chin% call_args) message("User-supplied 'header' will override that found in metadata.")
+      else header = as.logical(yaml_header$header)
+    }
+    if ('schema' %chin% yaml_names) {
+      new_types = sapply(yaml_header$schema$fields, `[[`, 'type')
+      if (any(null_idx <- sapply(new_types, is.null)))
+        new_types = do.call(c, new_types)
+      synonms = rbindlist(list(
+        character = list(syn = c('character', 'string')),
+        integer = list(syn = c('integer', 'int')),
+        numeric = list(syn = c('numeric', 'number', 'double'))
+      ), idcol = 'r_type')
```

Comment:
what about `factor = list(syn=c("factor","categorical"))` and `integer64`?

Reply:
Sure, missed those. Can keep an eye out for #2545 to expand it a bit more too.